graphql/src/store/prefetch.rs

```diff
@@ -567,66 +504,73 @@ fn execute_selection_set<'a>(
     resolver: &StoreResolver,
     ctx: &'a ExecutionContext<impl Resolver>,
     mut parents: Vec<Node>,
-    selection_sets: Vec<&'a q::SelectionSet>,
-    object_type: &ObjectOrInterface,
+    object_type: ObjectOrInterface,
+    grouped_field_set: IndexMap<&'a String, CollectedResponseKey<'a>>,
 ) -> Result<Vec<Node>, Vec<QueryExecutionError>> {
     let mut errors: Vec<QueryExecutionError> = Vec::new();
 
-    // Group fields with the same response key, so we can execute them together
-    let grouped_field_set = collect_fields(ctx, object_type, selection_sets, None);
+    if parents.is_empty() {
+        return Ok(Vec::new());
+    }
+
+    // Skip dead fragments as an optimization
+    let relevant_concrete_types: Option<HashSet<String>> = if parents[0].entity.is_empty() {
+        None
+    } else {
+        Some(parents.iter().map(|e| e.typename().to_owned()).collect())
+    };
 
     // Process all field groups in order
-    for (response_key, type_map) in grouped_field_set {
-        match ctx.deadline {
-            Some(deadline) if deadline < Instant::now() => {
+    for (response_key, collected_fields) in grouped_field_set {
+        if let Some(deadline) = ctx.deadline {
+            if deadline < Instant::now() {
                 errors.push(QueryExecutionError::Timeout);
                 break;
             }
-            _ => (),
         }
 
-        for (type_cond, fields) in type_map {
-            if !type_cond.matches(&parents) {
-                continue;
-            }
-
-            let concrete_type = type_cond
-                .matching_type(&ctx.query.schema.document, object_type)
-                .expect("collect_fields does not create type conditions for nonexistent types");
-
-            if let Some(ref field) = concrete_type.field(&fields[0].name) {
+        // Make sure the interface fields are processed first.
+        // See also: e0d6da3e-60cf-41a5-b83c-b60a7a766d4a
+        let interface_fields = collected_fields.interface_fields;
+        let interface = collected_fields
+            .interface_cond
+            .map(|cond| (ObjectOrInterface::Interface(cond), interface_fields));
+        for (type_cond, fields) in interface.into_iter().chain(
+            collected_fields
+                .concrete_types
+                .into_iter()
+                .filter(|(c, _)| match &relevant_concrete_types {
+                    None => true,
+                    Some(relevant_set) => relevant_set.contains(c.0.name.as_str()),
+                })
+                .map(|(c, f)| (ObjectOrInterface::Object(c.0), f)),
```

Comment:
There's a lot going on here, even though it's just about getting an iterator over all `(TypeCondition, Vec<Field>)`  pairs in the `collected_fields`. Could constructing the iterator be a function on `CollectedResponseKey`?

Reply:
I moved this into an `IntoIterator` implementation for `CollectedResponseKey`, now that we don't need the `relevant_set` anymore.